databroker/databroker.py

```diff
@@ -336,29 +343,43 @@ def get_events(headers, fields=None, fill=True):
     if fields is None:
         fields = []
     fields = set(fields)
-    for k in fields:
-        for header in headers:
-            for descriptor in header['descriptors']:
-                if k in descriptor['data_keys']:
-                    break
-            else:
-                raise ValueError(('{!r} field is not in any descriptor '
-                                  'of run start {rs.uid}').
-                                 format(k, header['run_star']))
+    _check_fields_exist(fields, headers)
 
     for header in headers:
-        for descriptor in header.descriptors:
-            all_fields = set(descriptor['data_keys'])
+        # cache these attribute look-ups for performance
+        start = header['start']
+        stop = header['stop']
+        if stop is None:
+            stop = {}
+        for descriptor in header['descriptors']:
+            objs_config = descriptor.get('configuration', {}).values()
+            config_data = merge(obj_conf['data'] for obj_conf in objs_config)
+            config_ts = merge(obj_conf['timestamps']
+                              for obj_conf in objs_config)
+            discard_fields = set()
+            extra_fields = set()
             if fields:
-                discard_fields = all_fields - fields
-            else:
-                discard_fields = []
-            if discard_fields == all_fields:
-                continue
+                event_fields = set(descriptor['data_keys'])
+                discard_fields = event_fields - fields
+                extra_fields = fields - event_fields
             for event in get_events_generator(descriptor):
+                event_data = event.data  # cache for perf
+                event_timestamps = event.timestamps
                 for field in discard_fields:
-                    del event.data[field]
-                    del event.timestamps[field]
+                    del event_data[field]
+                    del event_timestamps[field]
+                for field in extra_fields:
+                    # Look in the descriptor, then start, then stop.
+                    if field in config_data:
+                        event_data[field] = config_data[field]
+                        event_timestamps[field] = config_ts[field]
+                    elif field in start:
+                        event_data[field] = start[field]
+                        event_timestamps[field] = start['time']
+                    elif field in stop:
+                        event_data[field] = stop[field]
+                        event_timestamps[field] = stop['time']
```

Comment:
Wait no its not, since stop will be empty and field won't be in stop. Deleting this and the previous comment